### PR TITLE
Add support for impersonating a service account with BigQuery

### LINKED ISSUE #2672: Allow impersonation of Google Cloud service accounts with BigQuery
### Describe the feature
I recently read [this article](https://medium.com/@jryancanty/stop-downloading-google-cloud-service-account-keys-1811d44a97d9) about using the new IAM Credentials api in order to impersonate a service account by generating a  OAuth2 bearer tokens that identifies as them. Adding these abilities to dbt would help increase security and allow for additional abstraction when running BigQuery queries on sensitive data. This way, we don't need to run dbt with a service account key, or worry about starting a node identified by a particular service account.

This would allow us to invert control, using service accounts to control access to different types of data, and granting access to the service accounts (the "abstraction") instead of having all of the users have IAM access to the datasets directly. This way you get the benefits of federated identity with the flexibility of service account and their keys, which are typically required for this workflow, but have many downsides as described in the article.

From an end-user perspective, I propose that dbt adds an `impersonate_service_account` option to the profiles file for the BigQuery adapter. Using this functionality would work like this:
```yaml
dbt:
  target: dev
  outputs:
    dev:
      type: bigquery
      method: oauth
      project: myproject
      dataset: dbt_preston
      threads: 4
      timeout_seconds: 300
      location: US
      priority: interactive
      retries: 1
      impersonate_service_account: foobar@project.iam.gserviceaccount.com
```
In this case, the oauth/service account/service account file types would each be compatible. Instead of using the existing Credentials() directly as is the current functionality, when `impersonate_service_account` option is present, Credentials would be generated using the new IAM Credential API ([example](https://github.com/googleapis/google-auth-library-python/blob/15d5fa946177581b52a5a9eb3ca285c088f5c45d/google/auth/impersonated_credentials.py#L118)).

It seems like this would be pretty easy to send a patch for since the Google Auth library already has support for this method. There is one big difference that we may want to handle, these tokens are very short lived compared to ADC or Service Accounts. By default, these OAuth2 tokens expire after 1 hour. It is possible to increase this when requesting, but 1 hour is a soft limit, only able to be increased by adding the service account to an organization policy. This is usually only possible by GCP Org admins, so not your target audience. I think for an MVP it would be alright for this to be the fallback, but ideally there would be a way to detect the expired credentials and request a new set while dbt is running. This may already be possible, it's not clear how often new clients are requested. If a new client is used often, then this may not even be a concern since it would get a fresh token each time a job was created. In this case, we'd want to set the expiration much shorter than the default.

### Describe alternatives you've considered
The only other option is to have GCE or Cloud Run instances run as service accounts. Cloud Run times out after 15 minutes though currently so that is a non-starter, and GCE isn't really designed for this purpose. There are a bunch of other downsides as well that I'm happy to go into if the value isn't obvious.

### Additional context
This would be just for BigQuery users.

### Who will this benefit?
It will benefit BigQuery users with many datasets, each needing to be and remain isolated. Any operation concerned with increased security would prefer to use this, and most of us are.

### Are you interested in contributing this feature?
Yes, I am interested.

### PR DESCRIPTION
**Pull request by bbhoss**
resolves #2672

### Description

See #2672. Opening up this draft as I hope to get some early feedback. Tests and docs are the main TODOs, I still need to figure out how the testing works.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

**Comment by bbhoss** [comment]
I added a test but it doesn't seem very useful. Not really sure the existing tests are setup in a way to easily test this without major mocking, I don't see any other authentication tests. Let me know what else is needed.

**Comment by jtcohen6** [comment]
@bbhoss You're right, in general we don't have as much testing for non-default authentication methods (Snowflake SSO being the big exception). In this specific case, since CircleCI runs our integration tests against BigQuery using a service account (and it [doesn't sound like](https://circleci.com/docs/2.0/google-auth/#creating-and-storing-a-service-account) they support oauth), I don't know if we have a better option.

I'm still going to test this locally by oauth'ing into our sandboxed BigQuery account and impersonating Circle's service account.

**Comment by bbhoss** [comment]
@jtcohen6 added docs [here](https://github.com/fishtown-analytics/docs.getdbt.com/pull/323) in case you need a walkthrough for testing locally. Let me know if there's anything else I can do :)

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 ## dbt 0.18.0 (Release TBD)
 
+
+### Features
+- Add support for impersonating a service account using `impersonate_service_account` in the BigQuery profile configuration ([#2677](https://github.com/fishtown-analytics/dbt/issues/2677)) ([docs](https://docs.getdbt.com/reference/warehouse-profiles/bigquery-profile#service-account-impersonation))
+
 ### Breaking changes
 - `adapter_macro` is no longer a macro, instead it is a builtin context method. Any custom macros that intercepted it by going through `context['dbt']` will need to instead access it via `context['builtins']` ([#2302](https://github.com/fishtown-analytics/dbt/issues/2302), [#2673](https://github.com/fishtown-analytics/dbt/pull/2673))
```
Comment (jtcohen6) [inline CHANGELOG.md:9]: Glad I didn't forget—can you add a new "Contributors" section here (like the one below) and include your GitHub handle + this PR number? We want to give you credit!

**Pull request merged**